lc2fen/infer_pieces.py

```diff
@@ -164,26 +168,124 @@ def infer_chess_pieces(pieces_probs, a1_pos, previous_fen=None):
         If it is not None, improves piece inference.
     :return: A list of the inferred chess pieces in FEN notation order.
     """
+    if previous_fen is not None:
+        previous_board = chess.Board(previous_fen)
+        previous_list = board_to_list(fen_to_board(previous_fen))
     pieces_probs = board_to_list(list_to_board(pieces_probs, a1_pos))
 
     # None represents that no piece is set in that position yet
     out_preds = [None] * 64
 
     final_move_sq = -1
-    possible_pieces = []
-    if previous_fen is not None:
+    if previous_fen is not None:  # Perform move detection
         changed_squares_idx = changed_squares(previous_fen, pieces_probs)
         move = inferred_move(previous_fen, pieces_probs, changed_squares_idx)
-        if move is not None:
+        if (
+            move is not None
+        ):  # A move has been successfully detected so the FEN will be concluded immediately
             initial_sq, final_move_sq, action = move
-            possible_pieces = inferred_pieces_from_move(
-                initial_sq, final_move_sq, action
-            )
+            initial_coordinates = FILES[initial_sq % 8] + RANKS[initial_sq // 8]
+            final_coordinates = FILES[final_move_sq % 8] + RANKS[final_move_sq // 8]
+            move_UCI = initial_coordinates + final_coordinates
+            if action.startswith("white"):
+                previous_board.turn = chess.WHITE
+            else:
+                previous_board.turn = chess.BLACK
+            if (
+                previous_list[initial_sq] == "P" and initial_coordinates[1] == "7"
+            ):  # White promotes (and we have to figure out the promoted piece)
+                promoted_piece_prob = 0
+                if (
+                    pieces_probs[final_move_sq][4] > promoted_piece_prob
+                    and previous_fen.count("Q") < 2
```

Comment:
I just double-checked and there seems to be no limit to the promotions of pieces (i.e. you could have up to 9 pieces of each). This seems to be contrary also to what was changed in #12 and the comment in #13 . Is there any rationale behind this?
https://www.chess.com/terms/pawn-promotion
https://en.wikipedia.org/wiki/Promotion_(chess)

Reply:
Yeah I'm fully aware that theoretically we can have as many as 9 queens for either side. But this typically only happens in online chess where one player is "trolling" the other (one side would typically be left with a lone king or something while the other side would just delay delivering the checkmate as much as possible by getting more and more queens...).

A standard **physical** chess set provides at most two queens for either side (which, for all intents and purposes, is completely reasonable; with over 5 years of chess experience, I don't recall ever having to get 3 queens to win or draw in a game). Since this project is aimed at predicting FENs from images of **physical** chessboards, I think it makes sense to assume there are at most 2 queens on the board for either side

Reply:
This makes sense, it's probably a much better heuristic since those cases are quite niche.